url4fdf79ae2c_beckchr_juel_tgz-pJ8-de_odysseus_el_samples_xml_sax_XMELFilterJ8/src/de/odysseus/el/tree/impl/Scanner.java

```diff
@@ -0,0 +1,466 @@
+	 */
+	protected Scanner(String input) {
+		this.input = input;
+	}
+
+	public String getInput() {
+		return input;
+	}
+	
+	/**
+	 * @return current token
+	 */
+	public Token getToken() {
+		return token;
+	}
+
+	/**
+	 * @return current input position
+	 */
+	public int getPosition() {
+		return position;
+	}
+
+	/**
+	 * @return <code>true</code> iff the specified character is a digit
+	 */
+	protected boolean isDigit(char c) {
+		return c >= '0' && c <= '9';
+	}
+	
+	/**
+	 * @param s name
+	 * @return token for the given keyword or <code>null</code>
+	 */
+	protected Token keyword(String s) {
+		return KEYMAP.get(s);
+	}
+	
+	/**
+	 * @param symbol
+	 * @return token for the given symbol
+	 */
+	protected Token fixed(Symbol symbol) {
+		return FIXMAP.get(symbol);
+	}
+
+	protected Token token(Symbol symbol, String value, int length) {
+		return new Token(symbol, value, length);
+	}
+
+	protected boolean isEval() {
+		return token != null && token.getSymbol() != Symbol.TEXT && token.getSymbol() != Symbol.END_EVAL;
+	}
+	
+	/**
+	 * text token
+	 */
+	protected Token nextText() throws ScanException {
+		builder.setLength(0);
+		int i = position;
+		int l = input.length();
+		boolean escaped = false;
+		while (i < l) {
+			char c = input.charAt(i);
+			switch (c) {
+				case '\\':
+					if (escaped) {
+						builder.append('\\');
+					} else {
+						escaped = true;
+					}
+					break;
+				case '#':
+				case '$':
+					if (i+1 < l && input.charAt(i+1) == '{') {
+						if (escaped) {
+							builder.append(c);
+						} else {
+							return token(Symbol.TEXT, builder.toString(), i - position);
+						}
+					} else {
+						if (escaped) {
+							builder.append('\\');
+						}
+						builder.append(c);
+					}
+					escaped = false;
+					break;
+				default:
+					if (escaped) {
+						builder.append('\\');
+					}
+					builder.append(c);
+					escaped = false;
+			}
+			i++;
+		}
+		if (escaped) {
+			builder.append('\\');
+		}
+		return token(Symbol.TEXT, builder.toString(), i - position);
+	}
+	
+	/**
+	 * string token
+	 */
+	protected Token nextString() throws ScanException {
+		builder.setLength(0);
+		char quote = input.charAt(position);
+		int i = position+1;
+		int l = input.length();
+		while (i < l) {
+			char c = input.charAt(i++);
+			if (c == '\\') {
+				if (i == l) {
+					throw new ScanException(position, "unterminated string", quote + " or \\");
+				} else {
+					c = input.charAt(i++);
+					if (c == '\\' || c == quote) {
+						builder.append(c);
+					} else {
+						throw new ScanException(position, "invalid escape sequence \\" + c, "\\" + quote + " or \\\\");
+					}
+				}
+			} else if (c == quote) {
+				return token(Symbol.STRING, builder.toString(), i - position);
+			} else {
+				builder.append(c);
+			}
+		}
+		throw new ScanException(position, "unterminated string", String.valueOf(quote));
+	}
+	
+	/**
+	 * number token
+	 */
+	protected Token nextNumber() throws ScanException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.